File: src/book_workbench/settlement.py
"""Deterministic work-hour difference and transfer calculations.

This module deliberately does not contain language-model logic or spreadsheet
formatting. It accepts already-confirmed rows and returns auditable, typed
calculation results.
"""

from __future__ import annotations

from collections import defaultdict
from decimal import Decimal, InvalidOperation
from typing import Any, Iterable


class SettlementValidationError(ValueError):
    """Raised when confirmed settlement input is incomplete or invalid."""


def _decimal(value: Any, field: str, row_no: int | None = None) -> Decimal:
    location = f" on row {row_no}" if row_no is not None else ""
    if value is None or value == "":
        raise SettlementValidationError(f"{field} is required{location}")
    try:
        result = Decimal(str(value).strip())
    except (InvalidOperation, AttributeError):
        raise SettlementValidationError(f"{field} must be a number{location}") from None
    if not result.is_finite():
        raise SettlementValidationError(f"{field} must be finite{location}")
    return result


def _hours(value: Decimal) -> str:
    """Render Decimal hours without binary floating-point artefacts."""

    if value == value.to_integral_value():
        return str(value.quantize(Decimal("1")))
    return format(value.normalize(), "f")


def _name_key(name: str) -> str:
    return "".join(name.split())


def _status(difference: Decimal) -> str:
    if difference > 0:
        return "应收"
    if difference < 0:
        return "应转出"
    return "无差额"


def _normalise_rows(rows: Iterable[dict[str, Any]], issued_cap: Any) -> list[dict[str, Any]]:
    cap = _decimal(issued_cap, "issued_cap")
    if cap < 0:
        raise SettlementValidationError("issued_cap cannot be negative")

    normalised: list[dict[str, Any]] = []
    seen: set[str] = set()
    for row_no, row in enumerate(rows, start=1):
        if not isinstance(row, dict):
            raise SettlementValidationError(f"row {row_no} must be an object")
        raw_name = row.get("name")
        if raw_name is None or not str(raw_name).strip():
            raise SettlementValidationError(f"name is required on row {row_no}")
        name = str(raw_name).strip()
        key = _name_key(name)
        if key in seen:
            raise SettlementValidationError(f"duplicate person name: {name}")
        seen.add(key)

        actual = _decimal(row.get("actual_hours"), "actual_hours", row_no)
        issued = _decimal(row.get("issued_hours"), "issued_hours", row_no)
        if actual < 0:
            raise SettlementValidationError(f"actual_hours cannot be negative on row {row_no}")
        if issued < 0:
            raise SettlementValidationError(f"issued_hours cannot be negative on row {row_no}")
        if issued > cap:
            raise SettlementValidationError(
                f"issued_hours for {name} is {_hours(issued)}, above the cap {_hours(cap)}"
            )

        difference = actual - issued
        normalised.append(
            {
                "person_id": str(row["person_id"]).strip() if row.get("person_id") else None,
                "name": name,
                "actual_hours": actual,
                "issued_hours": issued,
                "difference_hours": difference,
                "status": _status(difference),
            }
        )
    if not normalised:
        raise SettlementValidationError("at least one person is required")
    return normalised


def calculate_settlement(rows: Iterable[dict[str, Any]], issued_cap: Any = "33") -> dict[str, Any]:
    """Calculate differences and a stable greedy transfer draft.

    Positive difference means the person should receive a private adjustment.
    Negative difference means the person has received more official hours than
    their confirmed actual hours and is a payer in the private adjustment.

    The function never changes ``issued_hours`` and never forces an unbalanced
    case to zero. Any remaining amount is reported as an external adjustment.
    """

    people = _normalise_rows(rows, issued_cap)
    receivers = [
        {"name": row["name"], "remaining": row["difference_hours"]}
        for row in people
        if row["difference_hours"] > 0
    ]
    payers = [
        {"name": row["name"], "remaining": -row["difference_hours"]}
        for row in people
        if row["difference_hours"] < 0
    ]

    transfers: list[dict[str, str]] = []
    # First close exact equal differences one-to-one. This avoids splitting a
    # transfer when a payer and receiver happen to have the same amount.
    for receiver in receivers:
        if receiver["remaining"] <= 0:
            continue
        for payer in payers:
            if payer["remaining"] > 0 and payer["remaining"] == receiver["remaining"]:
                transfers.append(
                    {
                        "payer": payer["name"],
                        "receiver": receiver["name"],
                        "hours": _hours(receiver["remaining"]),
                    }
                )
                payer["remaining"] = Decimal("0")
                receiver["remaining"] = Decimal("0")
                break

    receivers = [row for row in receivers if row["remaining"] > 0]
    payers = [row for row in payers if row["remaining"] > 0]
    # Then use largest residuals first. It is deterministic and usually keeps
    # the number of split transfers low while remaining easy to audit.
    receivers.sort(key=lambda row: (-row["remaining"], row["name"]))
    payers.sort(key=lambda row: (-row["remaining"], row["name"]))
    receiver_index = 0
    payer_index = 0
    while receiver_index < len(receivers) and payer_index < len(payers):
        receiver = receivers[receiver_index]
        payer = payers[payer_index]
        amount = min(receiver["remaining"], payer["remaining"])
        if amount > 0:
            transfers.append(
                {
                    "payer": payer["name"],
                    "receiver": receiver["name"],
                    "hours": _hours(amount),
                }
            )
        receiver["remaining"] -= amount
        payer["remaining"] -= amount
        if receiver["remaining"] == 0:
            receiver_index += 1
        if payer["remaining"] == 0:
            payer_index += 1

    actual_total = sum((row["actual_hours"] for row in people), Decimal("0"))
    issued_total = sum((row["issued_hours"] for row in people), Decimal("0"))
    difference_total = actual_total - issued_total
    receivable_total = sum((row["difference_hours"] for row in people if row["difference_hours"] > 0), Decimal("0"))
    payable_total = sum((-row["difference_hours"] for row in people if row["difference_hours"] < 0), Decimal("0"))
    transfer_total = sum((_decimal(item["hours"], "transfer hours") for item in transfers), Decimal("0"))

    return {
        "people": [
            {
                **row,
                "actual_hours": _hours(row["actual_hours"]),
                "issued_hours": _hours(row["issued_hours"]),
                "difference_hours": _hours(row["difference_hours"]),
            }
            for row in people
        ],
        "totals": {
            "actual_hours": _hours(actual_total),
            "issued_hours": _hours(issued_total),
            "difference_hours": _hours(difference_total),
            "receivable_hours": _hours(receivable_total),
            "payable_hours": _hours(payable_total),
            "draft_transfer_hours": _hours(transfer_total),
            "external_adjustment_hours": _hours(difference_total),
        },
        "transfers": transfers,
        "checks": {
            "issued_hours_within_cap": True,
            "people_unique": True,
            "difference_reconciles": actual_total - issued_total == difference_total,
            "draft_transfer_balanced_between_people": transfer_total == min(receivable_total, payable_total),
            "net_difference_is_zero": difference_total == 0,
        },
    }


def validate_transfers(
    rows: Iterable[dict[str, Any]],
    transfers: Iterable[dict[str, Any]],
    issued_cap: Any = "33",
) -> dict[str, Any]:
    """Validate an edited transfer draft against confirmed person differences."""

    normalised = _normalise_rows(rows, issued_cap)
    by_name = {_name_key(row["name"]): row for row in normalised}
    paid = defaultdict(lambda: Decimal("0"))
    received = defaultdict(lambda: Decimal("0"))
    errors: list[str] = []

    for row_no, transfer in enumerate(transfers, start=1):
        if not isinstance(transfer, dict):
            errors.append(f"transfer row {row_no} must be an object")
            continue
        payer_name = str(transfer.get("payer", "")).strip()
        receiver_name = str(transfer.get("receiver", "")).strip()
        try:
            amount = _decimal(transfer.get("hours"), "transfer hours", row_no)
        except SettlementValidationError as exc:
            errors.append(str(exc))
            continue
        payer = by_name.get(_name_key(payer_name))
        receiver = by_name.get(_name_key(receiver_name))
        if payer is None:
            errors.append(f"unknown payer: {payer_name}")
        if receiver is None:
            errors.append(f"unknown receiver: {receiver_name}")
        if amount <= 0:
            errors.append(f"transfer hours must be positive on row {row_no}")
        if payer_name and receiver_name and _name_key(payer_name) == _name_key(receiver_name):
            errors.append(f"payer and receiver cannot be the same person on row {row_no}")
        if payer is not None:
            if payer["difference_hours"] >= 0:
                errors.append(f"payer {payer_name} does not have a negative difference")
            paid[_name_key(payer_name)] += amount
        if receiver is not None:
            if receiver["difference_hours"] <= 0:
                errors.append(f"receiver {receiver_name} does not have a positive difference")
            received[_name_key(receiver_name)] += amount

    for key, row in by_name.items():
        allowed_paid = max(-row["difference_hours"], Decimal("0"))
        allowed_received = max(row["difference_hours"], Decimal("0"))
        if paid[key] > allowed_paid:
            errors.append(f"payer {row['name']} is over-allocated")
        if received[key] > allowed_received:
            errors.append(f"receiver {row['name']} is over-allocated")

    difference_total = sum((row["difference_hours"] for row in normalised), Decimal("0"))
    if difference_total != 0:
        errors.append(
            f"overall difference is {_hours(difference_total)}; adjust official issued hours before final transfer"
        )
    else:
        for key, row in by_name.items():
            required_paid = max(-row["difference_hours"], Decimal("0"))
            required_received = max(row["difference_hours"], Decimal("0"))
            if paid[key] != required_paid:
                errors.append(f"payer {row['name']} is not fully allocated")
            if received[key] != required_received:
                errors.append(f"receiver {row['name']} is not fully allocated")

    total_paid = sum(paid.values(), Decimal("0"))
    total_received = sum(received.values(), Decimal("0"))
    return {
        "ok": not errors and total_paid == total_received,
        "errors": errors if total_paid == total_received else [*errors, "paid and received totals do not match"],
        "paid_hours": _hours(total_paid),
        "received_hours": _hours(total_received),
        "unmatched_difference_hours": _hours(total_received - total_paid),
    }

File: settlement.js
// settlement.js — Deterministic work-hour difference and transfer calculations.
// Direct port of src/book_workbench/settlement.py. Uses decimal arithmetic via
// bigint to avoid float artefacts in hour totals.

const Settlement = (() => {
  // Represent hours internally as integer hundredths (e.g. 0.5h = 50).
  function _toHundredths(value, field, rowNo) {
    if (value == null || value === "") {
      throw new Error(`${field} is required${rowNo != null ? ` on row ${rowNo}` : ""}`);
    }
    const s = String(value).trim();
    // accept "10", "10.5", "10.50", "0.5"
    if (!/^\d+(\.\d+)?$/.test(s)) {
      throw new Error(`${field} must be a number${rowNo != null ? ` on row ${rowNo}` : ""}`);
    }
    const [whole, frac = ""] = s.split(".");
    const padded = (frac + "00").slice(0, 2);
    return Number(whole) * 100 + Number(padded || 0);
  }

  function _fromHundredths(h) {
    if (h % 100 === 0) return String(h / 100);
    return (h / 100).toFixed(2).replace(/0+$/, "").replace(/\.$/, "");
  }

  function _nameKey(name) { return String(name).split(/\s+/).join(""); }

  function _status(diff) {
    if (diff > 0) return "应收";
    if (diff < 0) return "应转出";
    return "无差额";
  }

  function _normaliseRows(rows, issuedCap) {
    const cap = _toHundredths(issuedCap, "issued_cap");
    if (cap < 0) throw new Error("issued_cap cannot be negative");
    const out = [];
    const seen = new Set();
    for (let i = 0; i < rows.length; i++) {
      const row = rows[i];
      if (!row || typeof row !== "object") throw new Error(`row ${i + 1} must be an object`);
      const raw = row.name;
      if (raw == null || !String(raw).trim()) throw new Error(`name is required on row ${i + 1}`);
      const name = String(raw).trim();
      const key = _nameKey(name);
      if (seen.has(key)) throw new Error(`duplicate person name: ${name}`);
      seen.add(key);
      const actual = _toHundredths(row.actual_hours, "actual_hours", i + 1);
      const issued = _toHundredths(row.issued_hours, "issued_hours", i + 1);
      if (actual < 0) throw new Error(`actual_hours cannot be negative on row ${i + 1}`);
      if (issued < 0) throw new Error(`issued_hours cannot be negative on row ${i + 1}`);
      if (issued > cap) {
        throw new Error(`issued_hours for ${name} is ${_fromHundredths(issued)}, above the cap ${_fromHundredths(cap)}`);
      }
      const diff = actual - issued;
      out.push({
        person_id: row.person_id ? String(row.person_id).trim() : null,
        name,
        actual_hours: actual,
        issued_hours: issued,
        difference_hours: diff,
        status: _status(diff),
      });
    }
    if (!out.length) throw new Error("at least one person is required");
    return out;
  }

  function calculate(rows, issuedCap = "33") {
    const people = _normaliseRows(rows, issuedCap);
    const receivers = people.filter(r => r.difference_hours > 0).map(r => ({ name: r.name, remaining: r.difference_hours }));
    const payers = people.filter(r => r.difference_hours < 0).map(r => ({ name: r.name, remaining: -r.difference_hours }));
    const transfers = [];

    // exact matches first
    for (const receiver of receivers) {
      if (receiver.remaining <= 0) continue;
      for (const payer of payers) {
        if (payer.remaining > 0 && payer.remaining === receiver.remaining) {
          transfers.push({ payer: payer.name, receiver: receiver.name, hours: _fromHundredths(receiver.remaining) });
          payer.remaining = 0;
          receiver.remaining = 0;
          break;
        }
      }
    }
    const recv2 = receivers.filter(r => r.remaining > 0);
    const pay2 = payers.filter(p => p.remaining > 0);
    recv2.sort((a, b) => b.remaining - a.remaining || a.name.localeCompare(b.name));
    pay2.sort((a, b) => b.remaining - a.remaining || a.name.localeCompare(b.name));
    let ri = 0, pi = 0;
    while (ri < recv2.length && pi < pay2.length) {
      const r = recv2[ri], p = pay2[pi];
      const amount = Math.min(r.remaining, p.remaining);
      if (amount > 0) {
        transfers.push({ payer: p.name, receiver: r.name, hours: _fromHundredths(amount) });
      }
      r.remaining -= amount;
      p.remaining -= amount;
      if (r.remaining === 0) ri++;
      if (p.remaining === 0) pi++;
    }

    let actualTotal = 0, issuedTotal = 0, receivableTotal = 0, payableTotal = 0, transferTotal = 0;
    for (const p of people) {
      actualTotal += p.actual_hours;
      issuedTotal += p.issued_hours;
      if (p.difference_hours > 0) receivableTotal += p.difference_hours;
      if (p.difference_hours < 0) payableTotal += -p.difference_hours;
    }
    for (const t of transfers) transferTotal += _toHundredths(t.hours, "transfer hours");
    const diffTotal = actualTotal - issuedTotal;

    return {
      people: people.map(p => ({
        ...p,
        actual_hours: _fromHundredths(p.actual_hours),
        issued_hours: _fromHundredths(p.issued_hours),
        difference_hours: _fromHundredths(p.difference_hours),
      })),
      totals: {
        actual_hours: _fromHundredths(actualTotal),
        issued_hours: _fromHundredths(issuedTotal),
        difference_hours: _fromHundredths(diffTotal),
        receivable_hours: _fromHundredths(receivableTotal),
        payable_hours: _fromHundredths(payableTotal),
        draft_transfer_hours: _fromHundredths(transferTotal),
        external_adjustment_hours: _fromHundredths(diffTotal),
      },
      transfers,
      checks: {
        issued_hours_within_cap: true,
        people_unique: true,
        difference_reconciles: actualTotal - issuedTotal === diffTotal,
        draft_transfer_balanced_between_people: transferTotal === Math.min(receivableTotal, payableTotal),
        net_difference_is_zero: diffTotal === 0,
      },
    };
  }

  function validateTransfers(rows, transfers, issuedCap = "33") {
    const people = _normaliseRows(rows, issuedCap);
    const byName = {};
    for (const p of people) byName[_nameKey(p.name)] = p;
    const paid = {}, received = {};
    const errors = [];

    for (let i = 0; i < transfers.length; i++) {
      const t = transfers[i];
      if (!t || typeof t !== "object") {
        errors.push(`transfer row ${i + 1} must be an object`);
        continue;
      }
      const payerName = String(t.payer || "").trim();
      const receiverName = String(t.receiver || "").trim();
      let amount;
      try {
        amount = _toHundredths(t.hours, "transfer hours", i + 1);
      } catch (e) {
        errors.push(e.message);
        continue;
      }
      const payer = byName[_nameKey(payerName)];
      const receiver = byName[_nameKey(receiverName)];
      if (!payer) errors.push(`unknown payer: ${payerName}`);
      if (!receiver) errors.push(`unknown receiver: ${receiverName}`);
      if (amount <= 0) errors.push(`transfer hours must be positive on row ${i + 1}`);
      if (payerName && receiverName && _nameKey(payerName) === _nameKey(receiverName)) {
        errors.push(`payer and receiver cannot be the same person on row ${i + 1}`);
      }
      if (payer) {
        if (payer.difference_hours >= 0) errors.push(`payer ${payerName} does not have a negative difference`);
        paid[_nameKey(payerName)] = (paid[_nameKey(payerName)] || 0) + amount;
      }
      if (receiver) {
        if (receiver.difference_hours <= 0) errors.push(`receiver ${receiverName} does not have a positive difference`);
        received[_nameKey(receiverName)] = (received[_nameKey(receiverName)] || 0) + amount;
      }
    }

    for (const [key, row] of Object.entries(byName)) {
      const allowedPaid = Math.max(-row.difference_hours, 0);
      const allowedReceived = Math.max(row.difference_hours, 0);
      if ((paid[key] || 0) > allowedPaid) errors.push(`payer ${row.name} is over-allocated`);
      if ((received[key] || 0) > allowedReceived) errors.push(`receiver ${row.name} is over-allocated`);
    }

    const diffTotal = people.reduce((s, r) => s + r.difference_hours, 0);
    if (diffTotal !== 0) {
      errors.push(`overall difference is ${_fromHundredths(diffTotal)}; adjust official issued hours before final transfer`);
    } else {
      for (const [key, row] of Object.entries(byName)) {
        const requiredPaid = Math.max(-row.difference_hours, 0);
        const requiredReceived = Math.max(row.difference_hours, 0);
        if ((paid[key] || 0) !== requiredPaid) errors.push(`payer ${row.name} is not fully allocated`);
        if ((received[key] || 0) !== requiredReceived) errors.push(`receiver ${row.name} is not fully allocated`);
      }
    }

    const totalPaid = Object.values(paid).reduce((a, b) => a + b, 0);
    const totalReceived = Object.values(received).reduce((a, b) => a + b, 0);
    const errs = totalPaid === totalReceived ? errors : [...errors, "paid and received totals do not match"];
    return {
      ok: errs.length === 0 && totalPaid === totalReceived,
      errors: errs,
      paid_hours: _fromHundredths(totalPaid),
      received_hours: _fromHundredths(totalReceived),
      unmatched_difference_hours: _fromHundredths(totalReceived - totalPaid),
    };
  }

  return { calculate, validateTransfers };
})();
